### PR TITLE
libpeas, libpeas2: Fix Python loader

### LINKED ISSUE #146239: gnome.gedit: Error loading plugin 'externaltools'
### Describe the bug
`gedit`'s plugin "External tools" cannot be activated

### Steps To Reproduce
1. Open `gedit` preferences (☰ → Preferences)
2. On plugins tab, activate "External tools"

### Expected behavior
"External tools" plugin activated

### Screenshots
image

### Additional context
1. At least required `python38Packages.pygobject3` 

<details>
<summary>without `python38Packages.pygobject3`</summary>

```console
$ gedit

** (gedit:12595): WARNING **: 15:24:05.707: Error initializing Python Plugin Loader: PyGObject initialization failed
ImportError: could not import gobject (error was: ModuleNotFoundError("No module named 'gi'"))

** (gedit:12595): WARNING **: 15:24:05.707: Please check the installation of all the Python related packages required by libpeas and try again

(gedit:12595): libpeas-WARNING **: 15:24:05.707: Loader 'python3' is not a valid PeasPluginLoader instance
```
</details>

<details>
<summary>with `python38Packages.pygobject3`</summary>

```console
$ nix-shell -p python38Packages.pygobject3 --run gedit

** (gedit:12567): WARNING **: 15:22:12.313: Failed to load shared library '/nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/libgedit-40.0.so' referenced by the typelib: /nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/libgedit-40.0.so: cannot open shared object file: No such file or directory

** (gedit:12567): WARNING **: 15:22:12.314: Error importing plugin 'externaltools':
Traceback (most recent call last):
  File "/nix/store/32q6ryrb860sksdi1al5djg3pgcpq92l-python3-3.8.11/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/gedit/plugins/externaltools/__init__.py", line 23, in <module>
    from .appactivatable import AppActivatable
  File "/nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/gedit/plugins/externaltools/appactivatable.py", line 78, in <module>
    class AppActivatable(GObject.Object, Gedit.AppActivatable):
  File "/nix/store/v8yqq0yc2prapdnn23cw4kzsxr1rfvn8-python3.8-pygobject-3.40.1/lib/python3.8/site-packages/gi/module.py", line 207, in __getattr__
    wrapper = metaclass(name, bases, dict_)
  File "/nix/store/v8yqq0yc2prapdnn23cw4kzsxr1rfvn8-python3.8-pygobject-3.40.1/lib/python3.8/site-packages/gi/types.py", line 247, in __init__
    register_interface_info(cls.__info__.get_g_type())
TypeError: must be an interface

(gedit:12567): libpeas-WARNING **: 15:22:12.315: Error loading plugin 'externaltools'
```
</details>

2. Also, app seems trying to call `/nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/libgedit-40.0.so` instead of `/nix/store/msn3v1kmihvii0jfxcjzm6z5a1dgwn9s-gedit-40.1/lib/gedit/libgedit-40.0.so` but I don't know how to fix it
> `lib/libgedit-40.0.so` instead of
> `lib/gedit/libgedit-40.0.so`

### Notify maintainers
@hedning @jtojnar @dasj19 @maxeaubrey

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.13.13-zen1, NixOS, 21.05.4114.195d5816cdd (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.05.4114.195d5816cdd, nixos-unstable-21.11pre330462.2fa862644fc"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: gnome.gedit
```

### PR DESCRIPTION
**Pull request by ghost**
Closes  #146239

## Description of changes

As per the discussion in https://github.com/NixOS/nixpkgs/pull/261152, this fixes loading plugins using the Python loader by adding PyGObject to the Python path before it tries to load the library, similarly to how `nautilus-python` does it.

The code for the Python loader is similar in both versions with only the line numbers changing so it can apply to both.

I added `wrapGAppsHook` to `libpeas` as it ships a demo app (`peas-demo`) that already includes a Python plugin that fails to run as the GObject Introspection paths aren't set.

For the other apps that got rebuilt by `nixpkgs-review`:

(edit bobby285271: some of these do have python plugins, but for now we will focus on pkgs touched in this PR)

<details>

- [x] `astroid`: No Python plugins.
- [x] `boatswain`: No Python plugins.
- [x] `budgie.budgie-control-center`: Rev-dep of budgie-desktop, does not use Peas.
- [x] `budgie.budgie-desktop`: No Python plugins, this PR was already part of https://github.com/NixOS/nixpkgs/pull/261152 so it should still work.
- [x] `calls`: No Python plugins.
- [x] `cinnamon`.xviewer: No Python plugins.
- [x] `endeavour`: No Python plugins.
- [x] `entangle`: Only Python standard library and `gi` imports.
- [x] `gedit`: Only Python standard library and `gi` imports.
- [x] `gitg`: No Python plugins.
- [x] `gnome-builder`: No Python plugins.
- [x] `gnome.cheese`: No Python plugins.
- [x] `gnome.eog`: No Python plugins.
- [x] `gnome.geary`: No Python plugins.
- [x] `gnome.pomodoro`: No Python plugins.
- [x] `gnome.totem`: Only Python standard library and `gi` imports (`dbus-python` seems to not be used anymore?).
- [x] `liferea`: Only Python standard library and `gi` imports.
- [x] `midori`: No Python plugins.
- [x] `mate.eom`: No Python plugins.
- [x] `mate.pluma`: "Snippets" plugin can't find `six`, fixed on this PR.
- [x] `microdnf`: No Python plugins.
- [x] `pantheon`.elementary-code: No Python plugins.
- [x] `pitivi`: No Python plugins.
- [x] `rhythmbox`: Only Python standard library and `gi` imports, the Zeitgeist plugin fails to load if Zeitgeist isn't installed (?).
- [x] `tilix`: Rev-dep of gtkd, no Python plugins.
- [x] `vala-language-server`: Depends on `gnome-builder`, does not use Peas.
- [x] `valent`: No Python plugins.
- [x] `xed-editor`: Only Python standard library and `gi` imports.
- [x] `xplayer`: Only Python standard library and `gi` imports.

</details>

@bobby285271 @jtojnar

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by FedericoSchonborn** [comment]
Result of `nixpkgs-review pr 273060` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>50 packages built:</summary>
  <ul>
    <li>astroid</li>
    <li>boatswain</li>
    <li>budgie.budgie-control-center</li>
    <li>budgie.budgie-control-center.debug</li>
    <li>budgie.budgie-desktop</li>
    <li>budgie.budgie-desktop-with-plugins</li>
    <li>budgie.budgie-gsettings-overrides</li>
    <li>budgiePlugins.budgie-analogue-clock-applet</li>
    <li>budgiePlugins.budgie-user-indicator-redux</li>
    <li>calls</li>
    <li>calls.devdoc</li>
    <li>cinnamon.xviewer</li>
    <li>endeavour</li>
    <li>entangle</li>
    <li>gedit</li>
    <li>gedit.devdoc</li>
    <li>gitg</li>
    <li>gnome-builder</li>
    <li>gnome-builder.devdoc</li>
    <li>gnome.cheese</li>
    <li>gnome.cheese.devdoc</li>
    <li>gnome.cheese.man</li>
    <li>gnome.eog</li>
    <li>gnome.eog.dev</li>
    <li>gnome.eog.devdoc</li>
    <li>gnome.geary</li>
    <li>gnome.pomodoro</li>
    <li>gnome.totem</li>
    <li>gtkd</li>
    <li>libpeas</li>
    <li>libpeas.dev</li>
    <li>libpeas.devdoc</li>
    <li>libpeas2</li>
    <li>libpeas2.dev</li>
    <li>libpeas2.devdoc</li>
    <li>liferea</li>
    <li>mate.eom</li>
    <li>mate.pluma</li>
    <li>microdnf</li>
    <li>midori</li>
    <li>midori-unwrapped</li>
    <li>pantheon.elementary-code</li>
    <li>pitivi</li>
    <li>pragha</li>
    <li>rhythmbox</li>
    <li>tilix</li>
    <li>vala-language-server</li>
    <li>valent</li>
    <li>xed-editor</li>
    <li>xplayer</li>
  </ul>
</details>

File: pkgs/desktops/mate/pluma/default.nix [hunk 1]
```diff
@@ -47,6 +48,11 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  postFixup = ''
+    buildPythonPath ${python3.pkgs.six}
```
Comment (bobby285271) [inline pkgs/desktops/mate/pluma/default.nix:52]: Nice catch :+1: . I still got lots of `Couldn't find foreign struct converter for 'cairo.Context'` though, last time I saw `Gtk.cairo_*` usages I just added `python3.pkgs.pycairo`, not sure if that applies here, but anyway I can at least enable the plugin now

File: pkgs/desktops/gnome/core/totem/default.nix [hunk 2]
```diff
@@ -91,9 +91,7 @@ stdenv.mkDerivation rec {
     adwaita-icon-theme
     gnome-desktop
     gsettings-desktop-schemas
-    # for plug-ins
     python3Packages.pygobject3
-    python3Packages.dbus-python
```
Comment (bobby285271) [inline pkgs/desktops/gnome/core/totem/default.nix:96]: Yeah in xplayer packaging (as a fork of old totem) documented this as for dbusservice plugin but this plugin is likely gone in new totem these days :shrug: : https://github.com/GNOME/totem/commit/166e95afd34e389121a5050cce677fb6ae90ce8a

**Review by bobby285271 — COMMENTED** [review]
I feel like we probably can still explicitly preserve pygobject3 if the app itself or the built-in plugins themselves require it in a direct way, e.g. [even without plugins, I believe pitivi still needs pygobject3](https://github.com/GNOME/pitivi/blob/2023.03.0/pitivi/application.py#L24)

**Review by bobby285271 — APPROVED** [review]

**Comment by jtojnar** [comment]
Changes look okay to me, would be nice to get a bit more context in the commit messages so that it is accessible in the future git blame.

**Comment by bobby285271** [comment]
Oh I clicked too fast

**Comment by jtojnar** [comment]
Never mind then :woman_shrugging:

**Pull request merged**